### DISCUSSION TITLE
LUA supported?

### DISCUSSION THREAD

**Discussion by yyq2013aa** [microsoft/garnet#80] [Q&A] [answered]
Would like to know does it support LUA just like reids?

**Comment by mgravell** [comment]
> hich makes it impossible to use the Redis implementation of IDistributedCache in aspnet.core.

The relevant library (Microsoft.Extensions.Caching.StackExchangeRedis) has been updated to support Garnet (and other similar RESP servers) from v9 onwards.

**Comment by badrishc** [comment, marked as answer]
Lua support was added a while back. Garnet supports Lua at this point. Although we do not encourage its usage except for legacy applications, because of its poor performance.

**Reply by nightroman** [reply]
My naive attempt to try scripting as in https://stackexchange.github.io/StackExchange.Redis/Scripting results in this error

```
StackExchange.Redis.RedisServerException: ERR This instance has Lua scripting support disabled
```

What steps are needed in order to enable scripting?

**Reply by badrishc** [reply]
You need to run the server with --lua argument.

**Reply by nightroman** [reply]
Great, thank you!

I could not find this and yet another important option `--lua-transaction-mode` in the docs.

The docs now declare Lua support but do not tell how to enable it. Is this deliberate?
If not, I may help adding these two switches to the configuration page.

**Reply by badrishc** [reply]
It wasn't deliberate. We would love to see more such PRs from you - we are dependent on the community to clean up these rough edges in the documentation and website. Really appreciate your contributions! :)

**Reply by BrunoScheufler** [reply]
> Lua support was added a while back. Garnet supports Lua at this point. Although we do not encourage its usage except for legacy applications, because of its poor performance.

Just to clarify your statement regarding poor performance, are you comparing this to implementing custom transactions in Garnet or a completely different mechanism? I'd be very interested to hear if the slowness comes from having to interpret Lua or the locking required for transactionality 😄

**Reply by kevin-montrose** [reply]
Lua is slow compared to extending Garnet via modules, which admittedly is not always an option.

Overhead comes both from interpreting Lua and the cost of p/invokes.  While steps have been taken to minimize both, the gap vs pure-.NET will always exist.

**Comment by mgravell** [comment]
Might want to check a few docs; if I search "Garnet lua" the first hit I
get is https://microsoft.github.io/garnet/docs/welcome/compatibility, which
says it is not included.

**Reply by badrishc** [reply]
OK - fixed it here: https://github.com/microsoft/garnet/pull/794

**Comment by HuaXiaoYi** [comment]
I execute lua script - EVAL "return 'hello lua'" 0 - on garnet server then it hang , what's the issue

**Reply by badrishc** [reply]
cc @kevin-montrose

**Reply by kevin-montrose** [reply]
Reported here: https://github.com/microsoft/garnet/issues/1079
Will be fixed as part of: https://github.com/microsoft/garnet/pull/1075

Root cause is a bug in compilation checking after a prior refactor.  We keep the invalid Lua instance cached, and things get dicey when it's used later.  Until that fix lands, the temporary work around would be to re-establish the connection since the cache is tied to the session.
